### PR TITLE
[RDY] Visual mode put from '.' register, also visual block mode put

### PR DESCRIPTION
**Pull request by hardenedapple**
Currently working on issues #5781  and #5709

### CODE REVIEW

**Comment by justinmk** [comment]
@hardenedapple FYI, https://github.com/vim/vim/commit/f8eb9c51e5bbd10e59c9b1247f8f6c7f5b77ccd0 may be useful.

Certainly the tests you added will be great to have as well.

**Comment by hardenedapple** [comment]
@justinmk  Thanks, I tested that change but it leaves the cursor in the wrong place sometimes, leaves the text in the wrong place with virtualedit past the end of the line, and doesn't work with visual block mode.

I'll go mention that on the change you've referenced once I've finished working on my solution.

File: test/functional/normal/put_spec.lua [hunk 1]
```diff
@@ -1,19 +1,15 @@
...
 local eq = helpers.eq
 local execute = helpers.execute
 local curbuf_contents = helpers.curbuf_contents
 local funcs = helpers.funcs
-local curwinmeths = helpers.curwinmeths
 local dedent = helpers.dedent
-local setreg, getreg = funcs.setreg, funcs.getreg
+local getreg = funcs.getreg
 
 local function setup()
```
Comment (justinmk) [inline test/functional/normal/put_spec.lua]: Is there a reason for not using busted's `setup(function()...end)` (runs once per `describe()` in the current scope) or `before_each(function()...end)` (runs once per `it()` in the current scope) methods?

**Comment by hardenedapple** [comment]
I've gotten the tests & code to somewhere I think it's good.

There were a few compromises I made by not fixing some bugs that I think are minor and would require tying this functionality to other parts of the program.

I've left these bugs in the tests for now, under a `pending()` block and with explanations as to why they're not fixed in comments above the tests.

I expect you don't want tests for `wontfix` bugs cluttering up the test output, I'm leaving them there as the easiest way of explaining the problem for this review stage.

I'll either remove the tests, comment them out, or if you'd prefer, fix the bugs (again I'll stress they would require ugly solutions like putting special cases in almost unrelated parts of the code predicated on information transferred via global variables) when it comes to merging.

File: test/functional/normal/put_spec.lua [hunk 2]
```diff
@@ -0,0 +1,1012 @@
...
+
+local function map(func, tab)
+    local rettab = {}
+    for k, v in pairs(tab) do
+        rettab[k] = func(v)
+    end
+    return rettab
+end
+
+local function filter(filter_func, tab)
```
Comment (justinmk) [inline test/functional/normal/put_spec.lua]: this and map() could go in `test/helpers.lua`

File: test/functional/normal/put_spec.lua [hunk 3]
```diff
@@ -0,0 +1,1012 @@
...
+    local rettab = {}
+    for _, entry in pairs(tab) do
+        if not filter_func(entry) then
+            table.insert(rettab, entry)
+        end
+    end
+    return rettab
+end
+
+local function splitlines(string)
```
Comment (justinmk) [inline test/functional/normal/put_spec.lua]: Is this like the VimL `split()` function? If possible it would be better to use `helpers.funcs.split()` which returns a lua table. If we have ad-hoc functions like this in the tests they're difficult to follow, one must check the implementation instead of seeing immediately the familiar VimL function.

File: src/nvim/ops.c [hunk 4]
```diff
@@ -2637,12 +2637,59 @@ void do_put(int regname, yankreg_T *reg, int dir, long count, int flags)
...
+        }
+    } else {
+        (void)stuff_inserted(command_start_char, count, false);
+    }
+    // Putting the text is done later, so can't move the cursor to the next
+    // character.  Simulate it with motion commands after the insert.
+    if (flags & PUT_CURSEND) {
+        if (flags & PUT_LINE) { stuffReadbuff((char_u *)"j0");
+        } else {
+            // Ensure that we don't ring the bell from attempting to move into
```
Comment (justinmk) [inline src/nvim/ops.c]: 2-space indent :>

File: test/functional/normal/put_spec.lua [hunk 5]
```diff
@@ -0,0 +1,1012 @@
...
+            -- inserting with 'a' behaves, we could add in a special case
+            -- whereby we set a flag in do_put() and read it when entering
+            -- insert mode but this seems like way too much to fix such a minor
+            -- bug.
+            pending('should be the same after ".pu. and ".pu.u.',
+                    make_cursor_test('".pu.'))
+        end)
+        -- The difficulty here is: The method used to put the dot register into
+        -- the buffer is by putting keys into the readbuff, in order to avoid
+        -- autoindent I would have to either find a vimL way of avoiding it and
```
Comment (justinmk) [inline test/functional/normal/put_spec.lua]: An ugly, but perhaps isolated, way to do it in viml might be `:set paste ..... :set nopaste`. (Let's keep that in a separate commit if you try it)

File: src/nvim/ops.c [hunk 6]
```diff
@@ -2637,12 +2637,59 @@ void do_put(int regname, yankreg_T *reg, int dir, long count, int flags)
...
+            // the space after the current line.
+            // We can stuff the readbuffer with "l" if
+            // 1) 'virtualedit' is "all" or "onemore"
+            // 2) We are not at the end of the line
+            // 3) We are not  (one past the end of the line && on the last line)
+            // The last condition is to account for a visual paste over a
```
Comment (justinmk) [inline src/nvim/ops.c]: I believe vim parlance is "put", not "paste". ("paste" is for clipboard stuff)

File: test/functional/normal/put_spec.lua [hunk 7]
```diff
@@ -0,0 +1,1012 @@
...
+            -- where the first character was inserted, not where the cursor was
+            -- when the 'a' was pressed.
+            -- We account for this the first time by saving the cursor position
+            -- in do_put(), but this isn't stored in redobuff for a second time
+            -- around.
+            -- We can't change how such a fundamental action as undo after
+            -- inserting with 'a' behaves, we could add in a special case
+            -- whereby we set a flag in do_put() and read it when entering
+            -- insert mode but this seems like way too much to fix such a minor
+            -- bug.
```
Comment (justinmk) [inline test/functional/normal/put_spec.lua]: Agreed, not worth a proliferating special case.

File: test/functional/normal/put_spec.lua [hunk 8]
```diff
@@ -0,0 +1,1012 @@
...
+            feed('".pu')
+            eq(origpos, funcs.getcurpos())
+        end)
+    end)
+
+    describe('register . problems', function()
+        before_each(reset)
+        -- I don't think these are reasonable to fix -- usually because the fix
+        -- adds too much interdependence for what it's worth.
+        -- I'm leaving them here for documentation.
```
Comment (justinmk) [inline test/functional/normal/put_spec.lua]: For the cases we don't plan to fix, I suggest pasting this entire `describe('register . problems' ...` source into the commit message.

File: test/functional/normal/put_spec.lua [hunk 9]
```diff
@@ -0,0 +1,1012 @@
...
+            	Line of words 1
+            Line of words 2]])
+            feed('u1go<C-v>j".p')
+            eq([[
+	ine of words 1
+	ine of words 2]], curbuf_contents())
+        end)
+
+        it('should not ring the bell with gp at end of line', function()
+            local Screen = require('test.functional.ui.screen')
```
Comment (justinmk) [inline test/functional/normal/put_spec.lua]: 2-space indent

File: test/functional/normal/put_spec.lua [hunk 10]
```diff
@@ -0,0 +1,1012 @@
...
+            local screen = Screen.new()
+            screen:attach()
+            helpers.ok(not screen.bell and not screen.visualbell)
+            execute('set beloff=""')
```
Comment (justinmk) [inline test/functional/normal/put_spec.lua]: Option name is misspelled. Errors in `helpers.execute()` are not bubbled up as lua errors. Each of the core functions at `test/functional/helpers.lua` have are documented to help avoid confusion about this. Basically the rule is: `feed()` (`nvim_input()`) never causes a client error, because it simulates user input. `execute()` happens to use feed(). This is admittedly confusing, I'd like to clean it up a bit...

File: src/nvim/ops.c [hunk 11]
```diff
@@ -2637,12 +2637,59 @@ void do_put(int regname, yankreg_T *reg, int dir, long count, int flags)
...
+            // selection one past the end of the line joining the current line
+            // with the one below.
+            colnr_T line_len = (colnr_T)STRLEN(ml_get(curwin->w_cursor.lnum));
```
Comment (justinmk) [inline src/nvim/ops.c]: Is this going to work on a multibyte line? Is `mb_string2cells()` more appropriate? 

May need to add a test with multibyte chars.

File: src/nvim/ops.c (hunk 4 above)
Comment (justinmk) [inline src/nvim/ops.c]: Could the logic in this block be factored out into a function? This decision about whether it's ok to nudge the cursor, is not uncommon. E.g. `cursor_can_nudge_forward()`

File: test/functional/normal/put_spec.lua [hunk 12]
```diff
@@ -0,0 +1,1012 @@
...
+    end)
+
+    describe('Visual put', function()
+        describe('basic put', function()
+            local test_string = [[
+            test_stringx" words 1
+            Line of words 2]]
+            run_normal_mode_tests(test_string, 'v2ep', nil, nil, 'Line of')
+        end)
+        describe('over trailing newline', function()
```
Comment (justinmk) [inline test/functional/normal/put_spec.lua]: These nested describe() cases should be it() cases. describe() is only for grouping.

```lua
describe('Visual put'', function()
  it('basic put', function()
  ...
  end)
  it('over trailing newline', function()
  ...
  end)

File: test/functional/normal/put_spec.lua (hunk 3 above)
Comment (hardenedapple) [inline test/functional/normal/put_spec.lua]: I originally used `funcs.split()`, but I wanted to do string processing before the connection to neovim session was made (i.e. before the tests were being run), and I get the error below in that case.
`./test/functional/helpers.lua:83: attempt to index upvalue 'session' (a nil value)`

I'm happy to move some function calls into a closure and delay splitting lines until the session has been connected if you'd prefer, but I thought I'd mention the downside before doing that.

File: test/functional/normal/put_spec.lua (hunk 12 above)
Comment (hardenedapple) [inline test/functional/normal/put_spec.lua]: I use `describe()` because the `run_test_variations()` function I use (through `run_normal_mode_tests()` in this case) calls `it()` for each of the generated test definitions i.e. `describe('basic put' ...)` groups the set of tests where each of `'p', 'gp', 'P', '".p', '2p', 'gP', ...` is tested in the basic case of putting over a visual selection.

Hence if I use an `it()` at the second level, I get the error `test/functional/normal/put_spec.lua:102: 'it' not supported inside current context block`.

That's also why I don't use `before_each()` very much: most of my tests are generated and run in `run_test_variations()`, so I thought I may as well put the setup function calls in that `it()` call.

It may be a little unclear (my bad), but it's how I make sure every put variation is tested in each situation.

File: src/nvim/ops.c (hunk 11 above)
Comment (hardenedapple) [inline src/nvim/ops.c]: I completely forgot to think about multibyte lines -- I'll go check that out. Thanks.

I don't think we can just disable bells, as then the user doesn't get a bell when the `".` register contains a backspace in a position that `'backspace'` doesn't allow it. I'll add a test for this.

Comment (hardenedapple) [inline src/nvim/ops.c]: With what I had, I was missing the case when the last character is a multibyte character. 

I believe `curwin->w_cursor.col` represents the byte position of the cursor (which is said to be at the start of whatever character it's on). 
As the value of `curwin->w_cursor.col` never increases past the number of bytes in the line (when `'virtualedit'` is active and we're past the end of the line `curwin->w_cursor.coladd` increments instead), I can avoid the string length calculation by just checking if the character under the cursor, or the character one multibyte after the cursor is the `NUL` marking the end of the line. I'll do that now.

File: src/nvim/ops.c (hunk 4 above)
Comment (hardenedapple) [inline src/nvim/ops.c]: It's actually a rather special case. I'm checking whether the cursor will be able to nudge forwards after the visual put has been executed.

File: src/nvim/ops.c [hunk 13]
```diff
@@ -2637,12 +2637,71 @@ void do_put(int regname, yankreg_T *reg, int dir, long count, int flags)
...
-      stuffcharReadbuff('l');
+    bool non_linewise_vis = (VIsual_active && VIsual_mode != 'V');
+
+    // PUT_LINE has special handling below which means we use 'i' to start.
+    char command_start_char = non_linewise_vis ? 'c' :
+        (flags & PUT_LINE ? 'i' : (dir == FORWARD ? 'a' : 'i'));
+
+    // To avoid being the affect of 'autoindent' on linewise puts, we create a
+    // new line with `:put _`.
+    if (flags & PUT_LINE) { do_put('_', NULL, dir, 1, PUT_LINE); }
```
Comment (mhinz) [inline src/nvim/ops.c]: Never put condition and body on the same line.

File: src/nvim/ops.c [hunk 14]
```diff
@@ -2637,12 +2637,71 @@ void do_put(int regname, yankreg_T *reg, int dir, long count, int flags)
...
+
+    // If given a count when pasting linewise, we stuff the readbuf with the
+    // dot register 'count' times split by newlines.
+    if (flags & PUT_LINE) {
+      stuffcharReadbuff(command_start_char);
+      for (; count > 0; count--) {
+        (void)stuff_inserted(NUL, 1, count != 1);
+        if (count != 1) { stuffcharReadbuff('\n'); }
```
Comment (mhinz) [inline src/nvim/ops.c]: cond ↯ body

File: src/nvim/ops.c [hunk 15]
```diff
@@ -2637,12 +2637,71 @@ void do_put(int regname, yankreg_T *reg, int dir, long count, int flags)
...
+      }
+    } else {
+      (void)stuff_inserted(command_start_char, count, false);
+    }
+    // Putting the text is done later, so can't move the cursor to the next
+    // character.  Simulate it with motion commands after the insert.
+    if (flags & PUT_CURSEND) {
+      if (flags & PUT_LINE) { stuffReadbuff((char_u *)"j0");
```
Comment (mhinz) [inline src/nvim/ops.c]: cond ↯ body

File: src/nvim/ops.c [hunk 16]
```diff
@@ -2637,12 +2637,71 @@ void do_put(int regname, yankreg_T *reg, int dir, long count, int flags)
...
+        bool virtualedit_allows = (ve_flags == VE_ALL
+                                   || ve_flags == VE_ONEMORE);
+        bool end_of_file = (
+          (curbuf->b_ml.ml_line_count == curwin->w_cursor.lnum)
+          && one_past_line);
+        if (virtualedit_allows || !(end_of_line || end_of_file)) {
+          stuffcharReadbuff('l');
+        }
+      }
+    } else if (flags & PUT_LINE) { stuffReadbuff((char_u *)"g'["); }
```
Comment (mhinz) [inline src/nvim/ops.c]: cond ↯ body

File: src/nvim/ops.c [hunk 17]
```diff
@@ -2831,14 +2890,10 @@ void do_put(int regname, yankreg_T *reg, int dir, long count, int flags)
...
-        curwin->w_cursor.col += (*mb_ptr2len)(get_cursor_pos_ptr());
-      else if (c != TAB || ve_flags != VE_ALL)
-        ++curwin->w_cursor.col;
-      ++col;
-    } else
-      getvcol(curwin, &curwin->w_cursor, &col, NULL, &endcol2);
+      // move to start of next multi-byte character
+      curwin->w_cursor.col += (*mb_ptr2len)(get_cursor_pos_ptr());
+      col++;
+    } else { getvcol(curwin, &curwin->w_cursor, &col, NULL, &endcol2); }
```
Comment (mhinz) [inline src/nvim/ops.c]: cond ↯ body

File: test/functional/normal/put_spec.lua (hunk 3 above)
Comment (justinmk) [inline test/functional/normal/put_spec.lua]: Oh, darn. Well, just call `clear()` at the top of the file then (before any describe() block) to force a session.

File: src/nvim/ops.c [hunk 18]
```diff
@@ -2637,12 +2637,86 @@ void do_put(int regname, yankreg_T *reg, int dir, long count, int flags)
...
+        // 1) 'virtualedit' is "all" or "onemore"
+        // 2) We are not at the end of the line
+        // 3) We are not  (one past the end of the line && on the last line)
+        // The last condition is to account for a visual put over a
+        // selection one past the end of the line joining the current line
+        // with the one below.
+
+        // curwin->w_cursor.col marks the byte position of the cursor in the
+        // curront line.
+        // It increases up to a max of STRLEN(ml_get(curwin->w_cursor.lnum)),
```
Comment (justinmk) [inline src/nvim/ops.c]: mb_ptr2len

Comment (hardenedapple) [inline src/nvim/ops.c]: I'm not sure what you mean here, but I think the comment is correct -- `curwin->w_cursor.col` can never increase past `STRLEN(ml_get(curwin->w_cursor.lnum))`, and it only reaches this value if it's past the end of the line (e.g. `'virtualedit'` or in visual characterwise mode).

**Comment by justinmk** [comment]
Does this also close  #5781 ?

**Comment by hardenedapple** [comment]
Yeah good point -- should have included that in the commit message.

File: test/helpers.lua [hunk 19]
```diff
@@ -90,11 +90,31 @@ local function tmpname()
...
+        rettab[k] = func(v)
+    end
+    return rettab
+end
+
+local function filter(filter_func, tab)
+    local rettab = {}
+    for _, entry in pairs(tab) do
+        if filter_func(entry) then
+            table.insert(rettab, entry)
```
Comment (justinmk) [inline test/helpers.lua]: 2-space indent

File: src/nvim/ops.c [hunk 20]
```diff
@@ -2637,12 +2637,86 @@ void do_put(int regname, yankreg_T *reg, int dir, long count, int flags)
...
+      for (; count > 0; count--) {
+        (void)stuff_inserted(NUL, 1, count != 1);
+        if (count != 1) {
+          // In order to avoid 'autoindent' affecting us if the text from the
+          // previous insert begins with whitespace, we use C-u to remove all
+          // whitespace at the beginning of the next line.
+          // We can't just insert Ctrl_U into readbuf1, because with
+          // 'backspace' including "eol" this would go back to the previous
+          // line if 'autoindent' is not on, hence we insert a dummy space just
+          // so that it can be removed with the C-u.
```
Comment (justinmk) [inline src/nvim/ops.c]: suggestion for brevity:

> To avoid 'autoindent' affecting the text, use Ctrl_U to remove any whitespace. Can't just insert Ctrl_U into readbuf1, this would go back to the previous line in the case of 'noautoindent' and 'backspace' includes "eol". So we insert a dummy space for Ctrl_U to consume.

File: src/nvim/ops.c [hunk 21]
```diff
@@ -2637,12 +2637,86 @@ void do_put(int regname, yankreg_T *reg, int dir, long count, int flags)
...
+      (flags & PUT_LINE ? 'i' : (dir == FORWARD ? 'a' : 'i'));
+
+    // To avoid being the affect of 'autoindent' on linewise puts, we create a
+    // new line with `:put _`.
+    if (flags & PUT_LINE) {
+      do_put('_', NULL, dir, 1, PUT_LINE);
+    }
+
+    // If given a count when pasting linewise, we stuff the readbuf with the
+    // dot register 'count' times split by newlines.
```
Comment (justinmk) [inline src/nvim/ops.c]: "putting" instead of "pasting"

File: src/nvim/ops.c [hunk 22]
```diff
@@ -2637,12 +2637,86 @@ void do_put(int regname, yankreg_T *reg, int dir, long count, int flags)
...
+        // The last condition is to account for a visual put over a
+        // selection one past the end of the line joining the current line
+        // with the one below.
+
+        // curwin->w_cursor.col marks the byte position of the cursor in the
+        // currunt line.
+        // It increases up to a max of STRLEN(ml_get(curwin->w_cursor.lnum)).
+        // When virtualedit is on, and the cursor is past the end of the line,
+        // curwin->w_cursor.coladd is incremented instead of
+        // curwin->w_cursor.col.
```
Comment (justinmk) [inline src/nvim/ops.c]:
```
        // Avoid ringing the bell from attempting to move into the space after
        // the current line. We can stuff the readbuffer with "l" if
        // 1) 'virtualedit' is "all" or "onemore"
        // 2) We are not at the end of the line
        // 3) We are not (one past the end of the line && on the last line).
        //    This accounts for a visual put over a selection one past the end
        //    of the line joining the current line with the one below.

        // curwin->w_cursor.col marks the byte position of the cursor in the
        // currunt line. It increases to a max of
        // STRLEN(ml_get(curwin->w_cursor.lnum)). With 'virtualedit' and the
        // cursor is past the end of the line, curwin->w_cursor.coladd is
        // incremented instead of curwin->w_cursor.col.
```

File: src/nvim/normal.c [hunk 23]
```diff
@@ -7727,16 +7724,23 @@ static void nv_put(cmdarg_T *cap)
...
-      cap->nchar = NUL;
-      cap->oap->regname = NUL;
-      nv_operator(cap);
-      do_pending_operator(cap, 0, false);
-      empty = (curbuf->b_ml.ml_flags & ML_EMPTY);
+      // In order to place the cursor in the correct position after pasting in
+      // blockwise visual mode, and to leave the text in the correct position
+      // when pasting with 'virtualedit' on and past the end of the line, we
+      // use the 'c' operator in do_put(), which requires the visual selection
+      // to still be active and to be highlighting the same text.
```
Comment (justinmk) [inline src/nvim/normal.c]:
```
        // To place the cursor correctly after a blockwise visual put, and to
        // leave the text in the correct position when putting with
        // 'virtualedit' and past the end of the line, we use the 'c' operator
        // in do_put(), which requires the visual selection to still be active.
```

**Comment by justinmk** [comment]
LGTM after final comments.

**Comment by justinmk** [comment]
`put_spec.lua` takes 30 seconds to run, it seems like each test takes 30-100 ms or more. 

https://s3.amazonaws.com/archive.travis-ci.org/jobs/192972619/log.txt

Is clear() or reset() being run more than once per test? Seems like something in the per-test setup is expensive.

File: test/helpers.lua (hunk 19 above)
Comment (hardenedapple) [inline test/helpers.lua]: oops ... you'd have thought I would have learned by now ... sorry!

**Comment by hardenedapple** [comment]
Yeah, that was exactly what was happening -- I thought a `before_each()` would not apply somewhere it did -- thanks for spotting that.

File: src/nvim/ops.c [hunk 24]
```diff
@@ -2637,12 +2637,81 @@ void do_put(int regname, yankreg_T *reg, int dir, long count, int flags)
...
+        if (!one_past_line) {
+          end_of_line = (*(cursor_pos + mb_ptr2len(cursor_pos)) == NUL);
+        }
+
+        bool virtualedit_allows = (ve_flags == VE_ALL
+                                   || ve_flags == VE_ONEMORE);
+        bool end_of_file = (
+            (curbuf->b_ml.ml_line_count == curwin->w_cursor.lnum)
+            && one_past_line);
+        if (virtualedit_allows || !(end_of_line || end_of_file)) {
```
Comment (justinmk) [inline src/nvim/ops.c]: For short-lived variables it's ok (and usually preferred) to have terse names. In the case of "end of file" and "end of line", abbreviations like `eol` and `eof` are often more familiar than the spell-out versions. I made this change in the merge.

**Comment by justinmk** [comment]
@hardenedapple Thanks!
